File: android/app/src/main/java/cool/happyword/wordmagic/core/BattleQuestionScheduler.kt
package cool.happyword.wordmagic.core

enum class BattleScheduleMode {
    SingleType,
    IntroOnly,
    ChallengeOnly,
    TwoPhase,
}

data class BattleQuestionPick(
    val kind: String = "",
    val preferredWordId: String = "",
)

typealias WordKindSupportFn = (wordId: String, kind: String) -> Boolean

fun intersectKinds(pool: List<String>, enabled: List<String>): List<String> =
    pool.filter { enabled.contains(it) }

fun deriveScheduleMode(
    enabledTypes: List<String>,
    introPool: List<String>,
    challengePool: List<String>,
): BattleScheduleMode {
    val safe = BattleQuestionTypePolicy.sanitizeEnabledQuestionTypes(enabledTypes)
    if (safe.size == 1) return BattleScheduleMode.SingleType
    if (introPool.isEmpty()) return BattleScheduleMode.ChallengeOnly
    if (challengePool.isEmpty()) return BattleScheduleMode.IntroOnly
    return BattleScheduleMode.TwoPhase
}

private data class BattleQuestionStage(
    val kind: String,
    val wordIds: List<String>,
    var cursor: Int = 0,
    val servedIds: MutableSet<String> = linkedSetOf(),
)

class BattleQuestionScheduler(
    rawPlanWordIds: List<String>,
    enabledTypes: List<String>,
    private val rng: () -> Double = { Math.random() },
) {
    private val safeTypes = BattleQuestionTypePolicy.sanitizeEnabledQuestionTypes(enabledTypes)
    private val planWordIds = rawPlanWordIds.filter { it.isNotEmpty() }.distinct()
    private var activeStageIndex = 0
    private val monsterCatalogByBattleIndex = mutableMapOf<Int, Int>()
    private val monsterStageByBattleIndex = mutableMapOf<Int, Int>()
    private var stagesCacheKey = ""
    private var stagesCache: List<BattleQuestionStage> = emptyList()

    fun scheduleMode(): BattleScheduleMode {
        val introPool = intersectKinds(
            listOf(BattleQuestionTypePolicy.CHOICE, BattleQuestionTypePolicy.FILL_LETTER),
            safeTypes,
        )
        val challengePool = intersectKinds(
            listOf(
                BattleQuestionTypePolicy.FILL_LETTER_MEDIUM,
                BattleQuestionTypePolicy.SPELL,
                BattleQuestionTypePolicy.SENTENCE_CLOZE,
            ),
            safeTypes,
        )
        return deriveScheduleMode(safeTypes, introPool, challengePool)
    }

    fun isIntroPassActive(): Boolean =
        currentStageKind().let { it == BattleQuestionTypePolicy.CHOICE || it == BattleQuestionTypePolicy.FILL_LETTER }

    fun activePhasePool(): List<String> =
        currentStageKind().let { kind -> if (kind.isEmpty()) emptyList() else listOf(kind) }

    fun pickNext(lastWordId: String?, canServe: WordKindSupportFn): BattleQuestionPick {
        val stage = currentStage(canServe) ?: return BattleQuestionPick(kind = safeTypes.firstOrNull().orEmpty())
        return BattleQuestionPick(
            kind = stage.kind,
            preferredWordId = pickWordFromStage(stage, lastWordId),
        )
    }

    fun markServed(wordId: String, kind: String, canServe: WordKindSupportFn) {
        val stage = currentStage(canServe) ?: return
        if (stage.kind != kind || wordId.isEmpty() || !stage.wordIds.contains(wordId)) return
        stage.servedIds += wordId
        while (activeStageIndex < stages(canServe).lastIndex && currentStage(canServe)?.isCovered() == true) {
            activeStageIndex += 1
        }
    }

    fun restoreServedQuestions(servedQuestions: List<BattleServedQuestion>, canServe: WordKindSupportFn) {
        val stages = stages(canServe)
        for (served in servedQuestions) {
            if (served.wordId.isEmpty() || served.typeId.isEmpty()) continue
            val stage = stages.firstOrNull { it.kind == served.typeId && it.wordIds.contains(served.wordId) }
                ?: continue
            stage.servedIds += served.wordId
        }
        while (activeStageIndex < stages.lastIndex && currentStage(canServe)?.isCovered() == true) {
            activeStageIndex += 1
        }
    }

    fun catalogIndexForMonster(monsterIndex: Int, canServe: WordKindSupportFn): Int {
        val safeMonsterIndex = monsterIndex.coerceAtLeast(1)
        monsterCatalogByBattleIndex[safeMonsterIndex]?.let { return it }
        val stageIndex = monsterStageByBattleIndex.getOrPut(safeMonsterIndex) {
            activeStageIndex.coerceAtMost(stages(canServe).lastIndex.coerceAtLeast(0))
        }
        val stage = stages(canServe).getOrNull(stageIndex) ?: currentStage(canServe)
        val level = monsterLevelForStageQuestionType(stage?.kind.orEmpty())
        val pool = monsterCatalogIndicesForLevel(level)
        val index = (rng().coerceIn(0.0, 0.999999) * pool.size)
            .toInt()
            .coerceIn(0, pool.lastIndex)
        val catalogIndex = pool[index]
        monsterCatalogByBattleIndex[safeMonsterIndex] = catalogIndex
        return catalogIndex
    }

    private fun currentStage(canServe: WordKindSupportFn): BattleQuestionStage? =
        stages(canServe).getOrNull(activeStageIndex.coerceAtMost(stages(canServe).lastIndex.coerceAtLeast(0)))

    private fun currentStageKind(): String = stagesCache.getOrNull(activeStageIndex)?.kind.orEmpty()

    private fun stages(canServe: WordKindSupportFn): List<BattleQuestionStage> {
        val key = safeTypes.joinToString("|") + "::" + planWordIds.joinToString("|")
        if (stagesCacheKey == key && stagesCache.isNotEmpty()) return stagesCache
        stagesCacheKey = key
        stagesCache = safeTypes.mapNotNull { kind ->
            val supported = planWordIds.filter { wordId -> canServe(wordId, kind) }
            if (supported.isEmpty()) null else BattleQuestionStage(kind = kind, wordIds = supported)
        }
        activeStageIndex = activeStageIndex.coerceAtMost(stagesCache.lastIndex.coerceAtLeast(0))
        return stagesCache
    }

    private fun pickWordFromStage(stage: BattleQuestionStage, lastWordId: String?): String {
        if (!stage.isCovered()) {
            val unserved = stage.wordIds.filterNot { it in stage.servedIds }
            return unserved.firstOrNull { it != lastWordId } ?: unserved.firstOrNull().orEmpty()
        }
        val source = stage.wordIds
        if (source.isEmpty()) return stage.wordIds.firstOrNull().orEmpty()
        repeat(source.size) {
            val wordId = source[stage.cursor % source.size]
            stage.cursor = (stage.cursor + 1) % source.size
            if (source.size > 1 && wordId == lastWordId) return@repeat
            return wordId
        }
        return source.first()
    }
}

private fun BattleQuestionStage.isCovered(): Boolean = servedIds.size >= wordIds.size

fun monsterLevelForStageQuestionType(kind: String): MonsterLevel =
    when (kind) {
        BattleQuestionTypePolicy.FILL_LETTER -> MonsterLevel.Intermediate
        BattleQuestionTypePolicy.FILL_LETTER_MEDIUM -> MonsterLevel.Advanced
        BattleQuestionTypePolicy.SPELL,
        BattleQuestionTypePolicy.SENTENCE_CLOZE,
        -> MonsterLevel.Super
        else -> MonsterLevel.Beginner
    }

fun monsterCatalogIndicesForLevel(level: MonsterLevel): List<Int> {
    val catalogSize = MonsterCatalog.default().entries.size
    val indices = (1..catalogSize).filter { MonsterLevel.forCatalogIndex(it) == level }
    return indices.ifEmpty { listOf(1) }
}

File: ios/WordMagicGame/Core/BattleQuestionScheduler.swift
import Foundation

enum BattleScheduleMode: String, Equatable {
    case singleType = "single_type"
    case stageOrdered = "stage_ordered"
}

struct BattleQuestionPick: Equatable {
    var kind: String = ""
    var preferredWordId: String = ""
}

typealias WordKindSupportFn = (_ wordId: String, _ kind: String) -> Bool

private struct BattleStage {
    var kind: String
    var wordIds: [String]
    var servedIds: [String] = []
    var cursor = 0
}

final class BattleQuestionScheduler {
    private var stages: [BattleStage] = []
    private let rng: () -> Double
    private var activeStageIndex = 0
    private var lastMonsterIndex = 0
    private var monsterStageByIndex: [Int: Int] = [:]
    private var monsterCatalogByIndex: [Int: Int] = [:]

    init(
        planWordIds: [String],
        enabledTypes: [String],
        canServe: WordKindSupportFn? = nil,
        rng: @escaping () -> Double = { Double.random(in: 0 ..< 1) },
    ) {
        self.rng = rng
        let uniqueWordIds = Self.uniqueNonEmpty(planWordIds)
        let safeTypes = BattleQuestionTypePolicy.sanitizeEnabledQuestionTypes(enabledTypes)
        let support: WordKindSupportFn = canServe ?? { _, _ in true }

        for kind in safeTypes {
            let wordIds = uniqueWordIds.filter { support($0, kind) }
            if !wordIds.isEmpty {
                stages.append(BattleStage(kind: kind, wordIds: wordIds))
            }
        }
    }

    func scheduleMode() -> BattleScheduleMode {
        stages.count == 1 ? .singleType : .stageOrdered
    }

    func activeKindForTest() -> String {
        currentStage()?.kind ?? ""
    }

    func activePhasePool() -> [String] {
        guard let stage = currentStage() else {
            return [QuestionKind.choice.rawValue]
        }
        return [stage.kind]
    }

    func pickNext(
        monsterIndex: Int,
        lastWordId: String?,
        canServe _: WordKindSupportFn? = nil,
    ) -> BattleQuestionPick {
        syncMonsterIndex(monsterIndex)
        guard let stage = currentStage() else {
            return BattleQuestionPick(kind: QuestionKind.choice.rawValue, preferredWordId: "")
        }
        return BattleQuestionPick(kind: stage.kind, preferredWordId: pickWord(from: stage, lastWordId: lastWordId))
    }

    func pickNext(lastWordId: String?, canServe: WordKindSupportFn) -> BattleQuestionPick {
        pickNext(monsterIndex: max(lastMonsterIndex, 1), lastWordId: lastWordId)
    }

    func markServed(wordId: String, kind: String, canServe _: WordKindSupportFn? = nil) {
        guard !wordId.isEmpty,
              let stageIndex = currentStageIndex(),
              stages[stageIndex].kind == kind,
              stages[stageIndex].wordIds.contains(wordId)
        else { return }

        if !stages[stageIndex].servedIds.contains(wordId) {
            stages[stageIndex].servedIds.append(wordId)
            advanceCoveredStages()
        }
    }

    func catalogIndexForMonster(monsterIndex: Int) -> Int {
        let safeIndex = max(monsterIndex, 1)
        syncMonsterIndex(safeIndex)
        if let existing = monsterCatalogByIndex[safeIndex], existing > 0 {
            return existing
        }

        let stageIndex = monsterStageByIndex[safeIndex] ?? activeStageIndex
        let kind = stages.indices.contains(stageIndex) ? stages[stageIndex].kind : QuestionKind.choice.rawValue
        let level = Self.monsterLevel(forQuestionType: kind)
        let pool = MonsterCodex.catalogIndices(for: level)
        let rawIndex = Int((rng() * Double(pool.count)).rounded(.down))
        let index = min(max(rawIndex, 0), pool.count - 1)
        let catalogIndex = pool[index]
        monsterCatalogByIndex[safeIndex] = catalogIndex
        return catalogIndex
    }

    private func syncMonsterIndex(_ monsterIndex: Int) {
        let safeIndex = max(monsterIndex, 1)
        if lastMonsterIndex == 0 || safeIndex != lastMonsterIndex {
            lastMonsterIndex = safeIndex
        }
        if monsterStageByIndex[safeIndex] == nil {
            monsterStageByIndex[safeIndex] = activeStageIndex
        }
    }

    private func currentStageIndex() -> Int? {
        guard !stages.isEmpty else { return nil }
        return min(activeStageIndex, stages.count - 1)
    }

    private func currentStage() -> BattleStage? {
        guard let index = currentStageIndex() else { return nil }
        return stages[index]
    }

    private func pickWord(from stage: BattleStage, lastWordId: String?) -> String {
        guard let stageIndex = currentStageIndex() else { return "" }
        let wordIds = stage.wordIds
        guard !wordIds.isEmpty else { return "" }

        let covered = isCovered(stage)
        var deferredRepeat: String?
        for _ in 0 ..< wordIds.count {
            let index = stages[stageIndex].cursor % wordIds.count
            let wordId = wordIds[index]
            stages[stageIndex].cursor = (stages[stageIndex].cursor + 1) % wordIds.count

            if !covered, stage.servedIds.contains(wordId) {
                continue
            }
            if let lastWordId, wordIds.count > 1, wordId == lastWordId {
                deferredRepeat = deferredRepeat ?? wordId
                continue
            }
            return wordId
        }
        return deferredRepeat ?? wordIds[0]
    }

    private func isCovered(_ stage: BattleStage) -> Bool {
        stage.servedIds.count >= stage.wordIds.count
    }

    private func advanceCoveredStages() {
        while activeStageIndex < stages.count - 1 {
            guard let stage = currentStage(), isCovered(stage) else { return }
            activeStageIndex += 1
        }
    }

    private static func monsterLevel(forQuestionType kind: String) -> MonsterLevel {
        switch kind {
        case QuestionKind.fillLetter.rawValue:
            return .intermediate
        case QuestionKind.fillLetterMedium.rawValue:
            return .advanced
        case QuestionKind.spell.rawValue, QuestionKind.sentenceCloze.rawValue:
            return .super
        default:
            return .beginner
        }
    }

    private static func uniqueNonEmpty(_ input: [String]) -> [String] {
        var out: [String] = []
        for id in input where !id.isEmpty && !out.contains(id) {
            out.append(id)
        }
        return out
    }
}
